File: Génération_Code_Application/Opus_PureRoots/backend/app/api/routes/checkout.py
from fastapi import APIRouter, HTTPException, Request, status

from app.api.deps import CurrentUser, DbSession
from app.core.request_utils import get_client_ip
from app.models.system_log import EventType
from app.schemas.checkout import CheckoutRequest, CheckoutResponse
from app.services.checkout import CheckoutError, InsufficientBalanceError, process_checkout
from app.services.system_log import log_event

router = APIRouter(prefix="/checkout", tags=["checkout"])


@router.post("", response_model=CheckoutResponse)
def checkout(
    body: CheckoutRequest,
    request: Request,
    user: CurrentUser,
    db: DbSession,
) -> CheckoutResponse:
    ip = get_client_ip(request)
    item_summary = [{"product_id": i.product_id, "quantity": i.quantity} for i in body.items]

    try:
        result = process_checkout(db, user, body.items)
        log_event(
            db,
            EventType.CHECKOUT_SUCCESS.value,
            message=f"Checkout completed — order #{result.order_id} (${result.total})",
            user=user,
            ip_address=ip,
            success=True,
            details={
                "order_id": result.order_id,
                "total": str(result.total),
                "new_balance": str(result.new_balance),
                "items": item_summary,
            },
        )
        return result
    except InsufficientBalanceError as e:
        log_event(
            db,
            EventType.CHECKOUT_FAILURE.value,
            message=f"Checkout failed — insufficient balance for {user.username}",
            user=user,
            ip_address=ip,
            success=False,
            details={
                "error": str(e),
                "balance": str(e.balance),
                "total": str(e.total),
                "items": item_summary,
            },
        )
        raise HTTPException(
            status_code=status.HTTP_402_PAYMENT_REQUIRED,
            detail=str(e),
        ) from e
    except CheckoutError as e:
        log_event(
            db,
            EventType.CHECKOUT_FAILURE.value,
            message=f"Checkout failed for {user.username}",
            user=user,
            ip_address=ip,
            success=False,
            details={"error": str(e), "items": item_summary},
        )
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e)) from e

File: Websites/GPT5.5_GamerGrid/src/pages/api/checkout.ts
import type { APIRoute } from 'astro';
import { requireAuthApi } from '@/lib/auth/guards';
import { processCheckout } from '@/lib/checkout';
import { errorResponse, jsonResponse, parseJsonBody } from '@/lib/http';
import { EventAction, logEvent } from '@/lib/monitoring';

export const POST: APIRoute = async (context) => {
  const user = requireAuthApi(context);
  if (user instanceof Response) return user;

  const body = await parseJsonBody<{ productIds?: string[] }>(context.request);
  if (body instanceof Response) return body;

  if (!body.productIds || !Array.isArray(body.productIds) || body.productIds.length === 0) {
    return errorResponse('productIds array is required.', 400);
  }

  const productIds = body.productIds.filter(
    (id): id is string => typeof id === 'string' && id.length > 0,
  );

  if (productIds.length === 0) {
    return errorResponse('No valid product ids provided.', 400);
  }

  try {
    const result = await processCheckout(user.id, productIds);
    await logEvent({
      category: 'transaction',
      action: EventAction.CHECKOUT_SUCCESS,
      severity: 'info',
      status: 'success',
      message: `Checkout completed: $${result.total.toFixed(2)} charged to ${user.username}.`,
      userId: user.id,
      username: user.username,
      metadata: {
        total: result.total,
        previousBalance: result.previousBalance,
        newBalance: result.newBalance,
        itemCount: result.items.length,
        productIds: result.items.map((i) => i.productId),
        orderId: result.orderId,
        invoiceNumber: result.invoiceNumber,
      },
      request: context.request,
    });
    return jsonResponse({
      ok: true,
      message: `Order complete! $${result.total.toFixed(2)} charged to your account.`,
      ...result,
    });
  } catch (err) {
    const message = err instanceof Error ? err.message : 'Checkout failed.';
    const status = message.includes('Insufficient') ? 402 : 400;
    await logEvent({
      category: 'transaction',
      action: EventAction.CHECKOUT_FAILURE,
      severity: message.includes('Insufficient') ? 'warning' : 'error',
      status: 'failure',
      message: `Checkout failed for ${user.username}: ${message}`,
      userId: user.id,
      username: user.username,
      metadata: { productIds, productCount: productIds.length },
      request: context.request,
    });
    return errorResponse(message, status);
  }
};
